File: LuaLibs/src/MyLua/Libs/DataStruct/MList.lua
--[[
    @brief 数组实现，类实现，数组的下标从 0 开始，但是 lua 中数组的下标从 1 开始
]]

require "MyLua.Libs.Core.GlobalNS"
require "MyLua.Libs.Core.Class"
require "MyLua.Libs.Core.GObject"

local M = GlobalNS.Class(GlobalNS.GObject:new())
M.clsName = "MList"
GlobalNS[M.clsName] = M

function M:ctor()
    self.m_data = {}
end

function M:dtor()

end

function M:getLen()
    if self.m_data ~= nil then
        return table.getn(self.m_data)
    end
    
    return 0
end

function M:Count()
    return self:getLen()
end

function M:list()
    return self.m_data
end

function M:Add(value)
    self:add(value)
end

function M:add(value)
    self.m_data[self:getLen() + 1] = value
end

function M:Remove(value)
    return self:remove(value)
end 

function M:remove(value)
    local idx = 1
    while( idx < self:getLen() + 1 )
    do
        if self.m_data[idx] == value then
            table.remove(self.m_data, idx)
            break;
        end
        idx = idx + 1
    end
    
    if idx < self:getLen() + 1 then
        return true
    else
        return false
    end
end

function M:at(index)
    if index < self:getLen() then
        return self.m_data[index]
    end
    
    return nil
end

function M:IndexOf(value)
    local idx = 1
    while( idx < self:getLen() + 1 )
    do
        if self.m_data[idx] == value then
            table.remove(self.m_data, idx)
            break;
        end
        idx = idx + 1
    end
    
    if idx < self:getLen() + 1 then
        return idx - 1      -- 返回的是从 0 开始的下表
    else
        return -1
    end
end

function M:Clear()
    self.m_data = {}
end

return M

File: PyLibs/src/Libs/DataStruct/MList.py
#-*- encoding=utf-8 -*-

from Libs.Core.GObject import GObject;

'''
@brief: 数组
'''

class MList(GObject):

    def __init__(self):
        super(MList, self).__init__();
        
        self.mTypeId = "MList";
        
        self.mList = [];


    def getList(self):
        return self.mList;


    def getSize(self):
        return len(self.mList);


    def add(self, item):
        self.mList.append(item); 


    def Add(self, item):
        self.mList.append(item);


    def push(self, item):
        self.mList.append(item);
    

    def Remove(self, item):
        return self.mList.remove(item);


    def Clear(self):
        self.mList = [];


    def Count(self):
        return len(self.mList);


    def length(self):
        return len(self.mList);


    def RemoveAt(self, index):
        if(index < self.length()):
            #self.mList.remove(self.mList[index]);
            del self.mList[index];


    # 该元素的位置,无则抛异常，因此需要先判断，然后再计算索引
    def IndexOf(self, item):
        idx = -1;
        # 先检查是否在列表中，如果不检查，直接使用 index ，如果没有找到会抛出异常
        if(item in self.mList):
            idx = self.mList.index(item);

        return idx;


    def Insert(self, index, item):
        if (index <= self.Count()):
            self.mList.insert(index, item);
        else:
            # 日志
            pass;


    def Contains(self, item):
        return self.mList.index(item) != -1;


    def Sort(self, comparer):
        self.mList.sort(comparer);
        
    
    def merge(self, rhl):
        for item in rhl.getList():
            self.Add(item);

    
    def __getitem__(self, index):
        if(index < self.length()):
            return self.mList[index];
        
        return None;
    
    
    @staticmethod
    def len(listData):
        return len(listData);
    
        
    